DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,308 @@
+
+##### "Accept interfaces, return structs"
+
+The Go community talks of the "accept interfaces, return structs" addage. The meaning of this idiom can be hard to pin down, herein lies some clarity.
+
+"Accept interfaces" speaks to the Dependency Inversion Principle (D in 'SOLID'). When a function declares dependent behavior as an interface parameter it gives control of _how_ that behavior is satisfied to the consuming logic. This is the foundation of flexible code.
+
+"return structs" is less obvious. It, generally, speaks to initializing constructors where authors should return the struct being constructed as opposed to interface it satisifes. This is largely due to duck typing (["If it walks like a duck and it quacks like a duck, then it must be a duck"](https://en.wikipedia.org/wiki/Duck_typing)).
+
+> Interfaces in Go provide a way to specify the behavior of an object: if something can do this, then it can be used here. - [Effective Go](https://go.dev/doc/effective_go#interfaces_and_types)
+
+Factory functions are not to be confused with intializing constructors. Factory functions typically return abstractions (an interface) satisfied by 1 of several implementations. A factory could be as simple as:
+
+```go
+// Foo and Bar satisfy fmt.Stringer
+func NewStringer(key string) (fmt.Stringer, error) {
+    switch key {
+    case "foo":
+        return Foo{}, nil
+    case "bar":
+        return Bar{}, nil
+    default:
+        return nil, errors.New("unknown key")
+    }
+}
+```
+
+Factories generally shouldn't contain complex construction logic. That's the job of the initializing constructor for the concrete type.
+
+##### Loggers
+
+Loggers are sometimes treated as a special depency by making entire projects depend on a package level logger. This _masks_ a crucial dependency that is hard to change without impacting the rest of the code base. To understand why this is bad, consider extracting a package into its own repository for reusability. Depending on a package level logger, or any other package level dependencies, makes the extraction process unnecessarily difficult.
+
+Logger's are not special dependencies, inject them like any other dependency. If you want to minimize the surface area of the default initializing constructor (`New...` funcs) you may construct noop loggers to satisfy the logger dependency and provide a means to override the logger with a setter.
+
+### Testing
+
+Software development is costly and hazardous. Tests provide example usage, guard rails, and assurances. Without tests, you cannot have high quality code. For this reason, all code we spend time designing and writing is worth testing. Tests should:
+
+- focus on the public API of a packge.
+- use the `_test` idiom for testing files.
+- treat the package as a black box.
+- cover at least 80% of the public API.
```

Comment:
clarification: testing 80% of the public functions/methods (eg. just calling those methods) or covering 80% of the code accessible through public APIs?

Reply:
Good point. 80% line coverage. Your code should be flexible enough to enable that. I've never had a problem and even strived for higher, 90%. With respect to the public API, we probably want something like 90%+ covered.

I think its worth calling out I chose this number based on personal experience. We don't have to stick to it but I'm definitely advocating for it.

Reply:
If this gets approved (which I do), let's talk about adding a rule to codecov.